src/connector/src/sink/sqlserver.rs

```diff
@@ -0,0 +1,649 @@
+// limitations under the License.
+
+use std::collections::{BTreeMap, HashMap};
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use risingwave_common::array::{Op, RowRef, StreamChunk};
+use risingwave_common::buffer::Bitmap;
+use risingwave_common::catalog::Schema;
+use risingwave_common::row::{OwnedRow, Row};
+use risingwave_common::types::{DataType, Decimal};
+use serde_derive::Deserialize;
+use serde_with::{serde_as, DisplayFromStr};
+use simd_json::prelude::ArrayTrait;
+use tiberius::numeric::Numeric;
+use tiberius::{AuthMethod, Client, ColumnData, Config, Query};
+use tokio::net::TcpStream;
+use tokio_util::compat::TokioAsyncWriteCompatExt;
+use with_options::WithOptions;
+
+use super::{SinkError, SINK_TYPE_APPEND_ONLY, SINK_TYPE_OPTION, SINK_TYPE_UPSERT};
+use crate::sink::writer::{LogSinkerOf, SinkWriter, SinkWriterExt};
+use crate::sink::{DummySinkCommitCoordinator, Result, Sink, SinkParam, SinkWriterParam};
+
+pub const SQLSERVER_SINK: &str = "sqlserver";
+
+fn default_max_batch_rows() -> usize {
+    1024
+}
+
+#[serde_as]
+#[derive(Clone, Debug, Deserialize, WithOptions)]
+pub struct SqlServerConfig {
+    #[serde(rename = "sqlserver.host")]
+    pub host: String,
+    #[serde(rename = "sqlserver.port")]
+    #[serde_as(as = "DisplayFromStr")]
+    pub port: u16,
+    #[serde(rename = "sqlserver.user")]
+    pub user: String,
+    #[serde(rename = "sqlserver.password")]
+    pub password: String,
+    #[serde(rename = "sqlserver.database")]
+    pub database: String,
+    #[serde(rename = "sqlserver.table")]
+    pub table: String,
+    #[serde(
+        rename = "sqlserver.max_batch_rows",
+        default = "default_max_batch_rows"
+    )]
+    #[serde_as(as = "DisplayFromStr")]
+    pub max_batch_rows: usize,
+    pub r#type: String, // accept "append-only" or "upsert"
+}
+
+impl SqlServerConfig {
+    pub fn from_btreemap(properties: BTreeMap<String, String>) -> Result<Self> {
+        let config =
+            serde_json::from_value::<SqlServerConfig>(serde_json::to_value(properties).unwrap())
+                .map_err(|e| SinkError::Config(anyhow!(e)))?;
+        if config.r#type != SINK_TYPE_APPEND_ONLY && config.r#type != SINK_TYPE_UPSERT {
+            return Err(SinkError::Config(anyhow!(
+                "`{}` must be {}, or {}",
+                SINK_TYPE_OPTION,
+                SINK_TYPE_APPEND_ONLY,
+                SINK_TYPE_UPSERT
+            )));
+        }
+        Ok(config)
+    }
+}
+
+#[derive(Debug)]
+pub struct SqlServerSink {
+    pub config: SqlServerConfig,
+    schema: Schema,
+    pk_indices: Vec<usize>,
+    is_append_only: bool,
+}
+
+impl SqlServerSink {
+    pub fn new(
+        mut config: SqlServerConfig,
+        schema: Schema,
+        pk_indices: Vec<usize>,
+        is_append_only: bool,
+    ) -> Result<Self> {
+        // Rewrite config because tiberius allows a maximum of 2100 params in one query request.
```

Comment:
Good catch! I wonder how you notice this corner case.

Reply:
found by a datagen table with high input rate.